### DISCUSSION TITLE
Brainstorming OVAL 6.0 and some type of conditional (if/else..) etc...

### DISCUSSION THREAD

**Discussion by vanderpol** [OVAL-Community/OVAL#170] [Ideas]
@OVAL-Community/oval-board-members 

We've had on occasion the need to exclude performing a certain test, because it may error out on that system (or for various other reasons).  Something like

If <Windows OS is a domain controller>
    Run some cmdlet test that we know is installed on domain controllers
else
   Run some other test that we know works for non-domain controllers

We currently have this 'working' by other means with all criteria AND/OR, but we have had to change some error/warnings to just debug when we come across a cmdlet that isn't installed and returns error, as all tests are run on all systems regardless.

This could be implemented like the XCCDF platform specification, or a true IF/ELSEIF/ELSE scenario.  

I'll admit I haven't put enough thought into this to make a full proposal and realize the consequences, but I suspect that I can't be the only person to dream of being able to do an IF in OVAL?  

IF there was a time (pun intended), OVAL 6 is it.  This is our one and only chance.   

IF you agree, and would like to see it happen, share your thoughts, proposals, etc...

**Comment by solind** [comment]
We've previously had discussions around making the applicability_check attribute of criteria/criterion/extend_definition consequential to the evaluation.  For the use-case you're describing, however, I've thought the XCCDF platform checks should suffice, since they can be used to gate whether OVAL checks should be evaluated on a target machine.

**Reply by vanderpol** [reply]
In our use case, this is all under a single DISA STIG XCCDF rule, which has two different methods for checking, dependent on the target.  As this has to tie back to a single XCCDF rule in the STIG manual, the logic of how to check falls to OVAL.  If we had 2 different rules, one for DC and one for Member Servers, XCCDF applicability would suffice.  We have used it in many other scenarios. 

If this is the only scenario, then we can just push DISA for a STIG update in the future, but I suspected that having some conditional checks in OVAL would be useful overall.

**Reply by solind** [reply]
My recollection of what's technically allowed is fuzzy, but, couldn't you have multiple oval checks with different platforms?  Then the evaluation phase could pick the right one to run.  Or... is the platform only specified at the rule/group level?

**Reply by vanderpol** [reply]
the xccdf platform is currently at the benchmark, group/rule level.  I think we would need to add it inside of the "check" for it to work the way we would need.  We can pursue that as well if we don't think that OVAL needs this capability.

**Reply by vojtapolasek** [reply]
> In our use case, this is all under a single DISA STIG XCCDF rule, which has two different methods for checking, dependent on the target. As this has to tie back to a single XCCDF rule in the STIG manual, the logic of how to check falls to OVAL. If we had 2 different rules, one for DC and one for Member Servers, XCCDF applicability would suffice. We have used it in many other scenarios.
> 
> If this is the only scenario, then we can just push DISA for a STIG update in the future, but I suspected that having some conditional checks in OVAL would be useful overall.

I think this could be expressed with XCCDF complex checks. I think you could check the XCCDF specification (attaching) section 6.4.4.4.
[xccdf.pdf](https://github.com/user-attachments/files/17254506/xccdf.pdf)

**Reply by vanderpol** [reply]
I might be mis-interpretting the XCCDF spec, but in my reading the XCCDF complex check still performs all of the checks called, it just allows the AND/OR to be calculated in XCCDF vs down in OVAL, primarily for mixing OVAL/OCIL etc.. From section 6.4.4.4 of the XCCDF spec:

"In place of an <xccdf:check> element, XCCDF allows an <xccdf:complex-check> element. An <xccdf:complex-check> is a boolean logical expression whose individual terms are <xccdf:check> and/or <xccdf:complex-check> elements."

**Comment by evgenyz** [comment]
Well, the state is definitely not ideal: https://github.com/OpenSCAP/openscap/issues/273. 

I would prefer a better not applicable/not implemented/not available states bubbling up from OVAL into XCCDF and better handling of these instead of having the same logic duplicated in OVAL.

**Reply by evgenyz** [reply]
I mean, if we are going to have SCAP 3.0 and OVAL 6.0, why can't we also have XCCDF 2.0?

**Reply by vanderpol** [reply]
Updates to XCCDF are definitely in-scope, as we have a few small changes submitted to @dragosprisaca at NIST.  If the changes would not break backward compatibility, they would fall into XCCDF 1.3.

**Reply by vanderpol** [reply]
In looking at your OpenSCAP issue, if you or anyone on your team would like to propose how OVAL 'should' work, this is the perfect time.  IMHO, there is some benefit in having duplicate logic in XCCDF as OVAL for conditional/applicability logic, as XCCDF can use applicability from OVAL (or OCIL) and then determine if the rule should be performed, even if that rule points to OCIL.  I realize nobody but our SCC team actually supports OCIL.   But I'd like to see what your team things OVAL should do, if it's already in the spec, and we just need to formalize it, maybe I'm overlooking something, but seems like something we should proceed with.

**Reply by maxullman** [reply]
Maybe this wouldn't be possible because it opens the door to circular references but an XCCDF change I would like to see is some ability to import evaluated variables from OVAL into XCCDF Values.

**Reply by evgenyz** [reply]
I guess your case pretty much boils down to inability of `criteria` system to choose a branch, ignoring test that is not available, but collecting some results from it (like why it is not available). In other words OVAL itself lacks proper NOT_APPLICABLE result in addition to ERROR, NOT_CHECKED etc.

This ideally should be all the author has to write:
```xml
      <criteria operator="OR">
        <criterion test_ref="oval:0:tst:platform_with_foo"/>
        <criterion test_ref="oval:0:tst:platform_without_foo"/>
      </criteria>
```

**Reply by maxullman** [reply]
I said OVAL into XCCDF but I think SCE into XCCDF would be more useful.  Some things are harder or impossible to express in OVAL so it's tempting to use SCE for those parts, but if you do that today, then you're stuck using SCE for the rest of the check too.  That use case is somewhat diminished with a shell test.

**Reply by vanderpol** [reply]
And what we're trying to accomplish is to figure out how to make OVAL easier to use, as SCE (not even sure what is is) isn't part of SCAP.  Now is the time to make OVAL better.

**Reply by solind** [reply]
@vanderpol  - SCE (which stands for Script Check Engine)  is just a way to embed a shell script in an SCAP benchmark (in ZIP bundle or XML datastream form), and map the exit code of a script execution to an XCCDF rule result.  It was invented by RedHat, and has been used pretty extensively by CIS.  And, since it can be embedded in a datastream, it can also be digitally signed.

The shell_command test is nice enough from a capabilities perspective, but sometimes you need a more complicated script to express the logic of a check.

If there's going to be a post-v1.3 SCAP, it could make sense to include it.

**Reply by vanderpol** [reply]
@solind, interesting to learn about SCE, especially after being shared a CIS benchmark that had their own commandshell OVAL test in it, which was also digitally signed.  The fact that we independently came up with essentially the same thing means SCAP/OVAL definitely needs something like it, one way or another.

I was hoping to have more of an official proposal, with more sample content, results and a demo build of SCC to share this week, but should have something by next week.  As FYI, the shellcommand test actually supports more than just a single command, anything you could pass into the shell interpreter, and we are expecting essentially shell scripts quite frequently, along with single commands, depending on what is needed.  We bounced around several ideas on what the OVAL test name should be, and are open to suggestions.

**Reply by vanderpol** [reply]
> ```
> platform_without_foo
> ```

And what I'd like to be able to do is to not process any specific test, based on the outcome of some other test.

If Test1 = True
     Process test 2
Else
    Process test 3

Tests 2 and 3 'may' or may not be applicable to the target, but I'd like to be able to not process ALL tests when I know that only a subset of them are needed to get the answer I'm looking for.   In my specific scenario, Test 2 happens to cause an error to be logged, because the cmdlet in question is not installed on the target, but the ability to do and IF/ELSE could be for any various reason.

**Reply by solind** [reply]
You know, you can determine if a cmdlet exists using the `Get-Command` cmdlet, passing in arguments `-errorAction SilentlyContinue` to avoid an error.  I bet you could write a cmdlet_test to check for the existence of your other cmdlet, and then structure the definition accordingly.

**Comment by wmunyan** [comment]
I know I'm not at CIS anymore and they've pretty much abandoned any new
implementation of SCAP/OVAL development, but wanted to offer my 2 cents of
historical context.  Back in the day, CIS-CAT implemented and proposed an
OVAL extension for "shellcommand", but the proposal to add that to the
official OVAL language was denied.  We moved forward with implementing SCE
to get around that, although probably left in the shellcommand stuff
because it was already there.  Unfortunately that opened Pandora's box for
CIS to abandon OVAL, drop all the people that had any OVAL experience or
advocated for it, and implement benchmarks using more and more SCE-based
checks.

I am sure there's probably some mailing list archives *somewhere* where the
discussions around the shellcommand_test could be found.

As a side note, I'd like to try and stay involved in the community here,
but obviously don't have a tool I can test any of these new proposals on,
but am really happy to see all the activity and renewed interest in the
next iterations of SCAP/XCCDF/OVAL.  Thanks for not purging me from the
mailing list :)

Cheers,
-Bill M.

**Reply by vanderpol** [reply]
Thanks for your background on this Bill, and I wish I had been more involved (or more open minded if by chance I was involved back when you proposed the OVAL shellcommand test.  You are very welcome to stay as part of this community, and we appreciate your insight.

**Reply by solind** [reply]
@wmunyan good to see you chiming in!  Hope all's well!

**Comment by vanderpol** [comment]
So my takeaway so far from this discussion is that OVAL lacks the ability to have conditional tests, but the general consensus seems to be that they are not needed, and there are ways to cobble around this limitation.   If there isn't enough interest in improving OVAL with this feature, we can call this discussion closed.  If I have misinterpreted the general consensus, please correct me, or add a proposal on how to improve OVAL.

**Reply by solind** [reply]
I believe that what I'm saying is that what OVAL "needs" is not so much conditional evaluation, what it needs is a more meaningful/useful way to specify applicability.  These are related, but not identical concepts.  And, I think your use-case would be better addressed by an improved applicability mechanism, than by the addition of a new conditional evaluation concept.

One of the familiar, even beneficial characteristics of OVAL is its blind obedience to collecting all the objects and evaluating all the tests, even when the outcome of a test is already knowable with what's been collected so far.  IMHO

**Reply by vanderpol** [reply]
I'm game for anything that gives us more granularity on what should be checked.  Any chance you have time to mock up what you think would scratch this itch?

**Reply by solind** [reply]
I previously had proposed a new "AND_APPLICABLE" operator that would have a new truth table, in which "not applicable" had weight.

But, like @wmunyan, I don't have access to an interpreter anymore.

**Reply by vanderpol** [reply]
If you still have a copy of the proposal, or where it might be (what year, etc..) to MITRE, or CIS?  I've searched online, and not finding it (which most of the MITRE managed proposals are easy to find online).  I've searched through the CIS email inbox archives, and didn't see anything with "AND_APPLICABLE".  If you have it or can point me to it, maybe someone on our team can prototype it

**Comment by vanderpol** [comment]
Closing this discussion for OVAL 6.0, we can revisit if/when OVAL 7.0 is created.
